### PR TITLE
[LRN] fix incorrect sign in output

### LINKED ISSUE #1559: [BUG][LRN][gfx90a] LRN output has incorrect sign
In some cases LRN output has incorrect sign. The simplest way to reproduce it is to modify [lrn_test.cpp](https://github.com/ROCmSoftwarePlatform/MIOpen/blob/develop/test/lrn_test.cpp) in the following way:
```diff
diff --git a/lrn_test.cpp b/lrn_test.cpp
index 084f0d3..48ae9f5 100644
--- a/lrn_test.cpp
+++ b/lrn_test.cpp
@@ -37,6 +37,7 @@
 #include <iterator>
 #include <limits>
 #include <iostream>
+#include <cstdio>

 template <class T>
 struct verify_lrn_foward
@@ -52,6 +53,13 @@ struct verify_lrn_foward

     tensor<T> cpu() const
     {
+        const float &ci0 = input(0, 0, 0, 0);
+        const float &ci1 = input(0, 1, 0, 0);
+        const float &ci2 = input(0, 2, 0, 0);
+        const_cast<float &>(ci0) = -3;
+        const_cast<float &>(ci1) = 0;
+        const_cast<float &>(ci2) = 3;
+
         auto output = tensor<T>{input.desc.GetLengths()};
         int n_batch, channels, height, width;
         std::tie(n_batch, channels, height, width) = miopen::tien<4>(input.desc.GetLengths());
@@ -108,11 +116,22 @@ struct verify_lrn_foward
             });
         }

+        float c0 = output(0, 0, 0, 0);
+        float c1 = output(0, 1, 0, 0);
+        float c2 = output(0, 2, 0, 0);
+        printf("c0:%f, c1:%f, c2:%f\n", c0, c1, c2); fflush(nullptr);
         return output;
     }

     tensor<T> gpu() const
     {
+        const float &ci0 = input(0, 0, 0, 0);
+        const float &ci1 = input(0, 1, 0, 0);
+        const float &ci2 = input(0, 2, 0, 0);
+        const_cast<float &>(ci0) = -3;
+        const_cast<float &>(ci1) = 0;
+        const_cast<float &>(ci2) = 3;
+
         auto&& handle = get_handle();
         auto out      = tensor<T>{input.desc.GetLengths()};
         auto in_dev   = handle.Write(input.data);
@@ -132,6 +151,11 @@ struct verify_lrn_foward
                     nullptr);

         out.data = handle.Read<T>(out_dev, out.data.size());
+        float c0 = out(0, 0, 0, 0);
+        float c1 = out(0, 1, 0, 0);
+        float c2 = out(0, 2, 0, 0);
+        printf("c0:%f, c1:%f, c2:%f\n", c0, c1, c2); fflush(nullptr);
+
         return out;
     }
```
I'm ussing the following command to run the test:
```bash
./test_lrn_test --float --mode Across_Channel --N 5 --alpha 0.000122 --beta 0.75 --K 1 --input 1 3 1 1
```
The output is the following:
```bash
c0:2.999012, c1:0.000000, c2:2.999012 # This is GPU
c0:-2.999012, c1:0.000000, c2:2.999012 # This is CPU
./bin/test_lrn_test --float --input 1, 3, 1, 1 --N 5 --alpha 0.000122 --beta 0.75 --K 1 --mode Across_Channel
FAILED: 1.1547
Iteration: 0
Input Tensor 1, 3, 1, 1
Max diff: 5.99802
Mismatch at 0: -2.99901 != 2.99901
```
As you can see the first value has incorrect sign. The LRN implementation for CPU in the test does comply with NVIDIA though MIOpen returns incorrect result.

Software versions:
- MIOpen 2.16 (HIP)
- ROCm 4.5

### PR DESCRIPTION
**Pull request by densamoilov**
This PR fixes #1559 and adjusts initialization of the input data in `MIOpenDriver` for LRN and `test_lrn_test` to test the cases when input contains negative values.

### CODE REVIEW

File: src/kernels/MIOpenLRNFwd.cl [hunk 1]
```diff
@@ -410,10 +410,12 @@ MIOpenLRNAcrossChannels4(const __global _FLOAT* bottom,
     int b               = get_global_id(2); // batch
     MLO_READ_TYPE accum = 0;
     MLO_READ_TYPE bot_in2[MLO_LRN_KERNEL_SZ];
+    MLO_READ_TYPE bot_in[MLO_LRN_KERNEL_SZ];
     int c_i = 0, c_o = 0;
     for(int i = 0; i < MLO_LRN_KERNEL_SZ; ++i)
     {
         bot_in2[i] = 0;
+        bot_in[i] = 0;
```
Comment (junliume) [inline src/kernels/MIOpenLRNFwd.cl:418]: Hi @densamoilov, first of all, thank you very much for raising this PR!
We have some issues with Clang Format Tidy with this line and several lines below, likely due to a tab in front of "=" sign.
The best way to get rid of the format issue is to follow procedure outlined here:
https://github.com/ROCmSoftwarePlatform/MIOpen/wiki/How-to-format-code 

Thanks!

Comment (junliume) [inline src/kernels/MIOpenLRNFwd.cl:418]: Hi @densamoilov, first of all, thank you very much for raising this PR!
We have some issues with Clang Format Tidy with this line and several lines below, likely due to a tab in front of "=" sign.
The best way to get rid of the format issue is to follow procedure outlined here:
https://github.com/ROCmSoftwarePlatform/MIOpen/wiki/How-to-format-code 

Thanks!

Comment (densamoilov) [reply src/kernels/MIOpenLRNFwd.cl]: Thanks for pointing this out! I fixed the format issues.

Comment (DrizztDoUrden) [reply src/kernels/MIOpenLRNFwd.cl]: You can use something like this to only format files changed from baseline if you want (this implies WSL, but is easily modified to be called from native linux if you want):
```Powershell
Param(
	[Parameter(ParameterSetName="Work")]
	[String]$Distro = "Ubuntu-20.04",
	[Parameter(ParameterSetName="Work")]
	[String]$ClangFormat = "clang-format-10",
	[Parameter(ParameterSetName="Work")]
	[Switch]$Log,
	[Parameter(ParameterSetName="Work")]
	[Switch]$SingleThread,
	[Parameter(ParameterSetName="List", Mandatory=$true)]
	[Switch]$List,
	[Switch]$All)

Import-Module PoshRSJob

$branch = git rev-parse --abbrev-ref HEAD
$files = @()

if ($All)
{
	$files += git ls-tree --full-tree -r --name-only HEAD
}
else
{
	$files += git diff --name-only remotes/origin/develop..$branch
	$files += git clean -dn | % { $_.Substring(13) }
}

$files = $files| ? { $_ -Match "\.h$|\.hpp$|\.c$|\.cpp$|\.h\.in$|\.hpp\.in$|\.cpp\.in$|\?CMakeLists.txt" }

if ($List)
{
	$files
	return
}

$repoRoot = git rev-parse --show-toplevel
$jobs = @()

$Worker = {
	Param($file, $repoRoot, $Log, $Distro, $ClangFormat)
	Set-Location $repoRoot
	$out = wsl --distribution "$Distro" --exec "$ClangFormat" -i -style=file "$file"
	if ($Log) { $out }
	if ($LASTEXITCODE -eq 0) { if ($Log) { "Finished formating `"$file`"." } }
	else { "Error formating `"$file`"." }
}

$jobId = 0

foreach ($file in $files)
{
	$jobId++

	if (-not (Test-Path -PathType Leaf $file))
	{
		if ($Log) { Write-Host -ForegroundColor Cyan "File doesn't exist: `"$file`"`r" }
		Write-Progress -Id 0 -Activity "clang-format" -CurrentOperation "Starting" -PercentComplete (100 * $jobId / $files.Count) -Status "File doesn't exist: `"$file`""
	}
	else
	{
		if ($Log) { Write-Host -ForegroundColor Cyan "wsl --distribution $Distro --exec $ClangFormat -i -style=file $file" }
		Write-Progress -Id 0 -Activity "clang-format" -CurrentOperation "Starting" -PercentComplete (100 * $jobId / $files.Count) -Status "clang-format-3.8 -i -style=file `"$file`""
		if ($SingleThread) { & $Worker $file $repoRoot $Log $Distro $ClangFormat }
		else { $jobs += Start-RSJob -ScriptBlock $Worker -ArgumentList $file, $repoRoot, $Log, $Distro, $ClangFormat }
	}
}

Write-Progress -Id 0 -Activity "clang-format" -Completed

Wait-RSJob -Job $jobs -ShowProgress | Out-Null
$jobs | Receive-RSJob;
$jobs | Remove-RSJob;
```

**Comment by atamazov** [comment]
@densamoilov 
> [force-pushed](https://github.com/ROCmSoftwarePlatform/MIOpen/compare/fb5d81d38fbfa08976fffd7277be31ead9a91db2..a60bfe0fbd9f1202fda7ab1a591edafe52cb55c2) the fix-lrn-output-sign branch from[ fb5d81d ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/fb5d81d38fbfa08976fffd7277be31ead9a91db2)to[ a60bfe0 ](https://github.com/ROCmSoftwarePlatform/MIOpen/commit/a60bfe0fbd9f1202fda7ab1a591edafe52cb55c2)5 days ago

Please do not **_force-push_**. This is NOT recommended due to several reasons. Force-pushing can be more or less safely used only if the branch has been published for review yet.

**Comment by densamoilov** [comment]
@junliume, is there any chance to backport this fix to 2.18?

**Comment by junliume** [comment]
> @junliume, is there any chance to backport this fix to 2.18?

MIOpen version has recently bumped to 2.18, so if this change is implemented it will be part of 2.18 eventually.
Do you mean backport to 2.17 and before? Thanks!

**Comment by densamoilov** [comment]
> Do you mean backport to 2.17 and before?

I must have mixed up the versions, my question was about 2.17 (the next release).

**Comment by junliume** [comment]
> > Do you mean backport to 2.17 and before?
> 
> I must have mixed up the versions, my question was about 2.17 (the next release).

The next release is coming up very soon, so maybe ROCm 5.3 is more likely.

P.S. would you add the test case? It's okay and I can create it as a separate issue and ask the team to do that.

**Comment by densamoilov** [comment]
> P.S. would you add the test case? It's okay and I can create it as a separate issue and ask the team to do that.

I actually adjusted filling input data so that the data has negative values. This triggers errors without the fix so I think the case is tested now.
I did it for [MIOpen driver](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1562/files#diff-330ec712406ccca98b9d10e0851061c3204ee056f77babd3104a4a630a850615R268) and for a [dedicated LRN test](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/1562/files#diff-9a8b509cc3432868647a9f79c16e71d95b7089e434f913c9127e988b08b3bcb6R288)

**Comment by junliume** [comment]
Ping @JehandadKhan @DrizztDoUrden @averinevg for review

**Review by junliume — APPROVED** [review]

**Review by JehandadKhan — APPROVED** [review]

**Pull request merged**